### ISSUE TITLE
Errors when changing match while an object is despawning

### ISSUE THREAD

**Issue by maxkratt** [FirstGearGames/FishNet#99]
Two errors sometimes appear when a connection changes matches (observer condition) 
while an object is despawning. After the errors the object sometimes enters a state where both
IsServer and IsClient return false. 

I've been able to consistently recreate the errors with this setup:
A spawnable object that despawns itself after a short while. And an object that changes matches and spawns the spawnable object.

Here is a unitypackage with the test files:
[ErrorWhenChangingMatch.zip](https://github.com/FirstGearGames/FishNet/files/9438693/ErrorWhenChangingMatch.zip)

To get the errors, I spawn some objects by pressing the 'S' key, and switch matches by tapping the
'A' key until the errors appear.

The test was performed on Unity 2020.3.33f1, FishNet 2.3.14.Pro, Server + Client in the Unity Editor.

These are the errors:
```
NullReferenceException: Object reference not set to an instance of an object
FishNet.Managing.Client.ClientObjectCache.Iterate () (at Assets/FishNet/Runtime/Managing/Client/Object/ObjectCaching.cs:314)
FishNet.Managing.Client.ClientObjects.IterateObjectCache () (at Assets/FishNet/Runtime/Managing/Client/Object/ClientObjects.cs:386)
FishNet.Managing.Client.ClientManager.ParseReceived (FishNet.Transporting.ClientReceivedDataArgs args) (at Assets/FishNet/Runtime/Managing/Client/ClientManager.cs:396)
FishNet.Managing.Client.ClientManager.Transport_OnClientReceivedData (FishNet.Transporting.ClientReceivedDataArgs args) (at Assets/FishNet/Runtime/Managing/Client/ClientManager.cs:231)
FishNet.Transporting.Tugboat.Tugboat.HandleClientReceivedDataArgs (FishNet.Transporting.ClientReceivedDataArgs receivedDataArgs) (at Assets/FishNet/Runtime/Transporting/Transports/Tugboat/Tugboat.cs:217)
FishNet.Transporting.Tugboat.Client.ClientSocket.IterateIncoming () (at Assets/FishNet/Runtime/Transporting/Transports/Tugboat/Core/ClientSocket.cs:288)
FishNet.Transporting.Tugboat.Tugboat.IterateIncoming (System.Boolean server) (at Assets/FishNet/Runtime/Transporting/Transports/Tugboat/Tugboat.cs:190)
FishNet.Managing.Transporting.TransportManager.IterateIncoming (System.Boolean server) (at Assets/FishNet/Runtime/Managing/Transporting/TransportManager.cs:364)
FishNet.Managing.Timing.TimeManager.TryIterateData (System.Boolean incoming) (at Assets/FishNet/Runtime/Managing/Timing/TimeManager.cs:955)
FishNet.Managing.Timing.TimeManager.IncreaseTick () (at Assets/FishNet/Runtime/Managing/Timing/TimeManager.cs:687)
FishNet.Managing.Timing.TimeManager.TickUpdate () (at Assets/FishNet/Runtime/Managing/Timing/TimeManager.cs:361)
FishNet.Transporting.NetworkReaderLoop.Update () (at Assets/FishNet/Runtime/Transporting/NetworkReaderLoop.cs:28)
```

```
NetworkObject for ObjectId of 2 was found null. Unable to despawn object.
UnityEngine.Debug:LogError (object)
FishNet.Managing.Client.ClientObjectCache:<Iterate>g__ProcessObject|11_0 (FishNet.Managing.Client.CachedNetworkObject,bool,int,FishNet.Managing.Client.ClientObjectCache/<>c__DisplayClass11_0&) (at Assets/FishNet/Runtime/Managing/Client/Object/ObjectCaching.cs:233)
FishNet.Managing.Client.ClientObjectCache:Iterate () (at Assets/FishNet/Runtime/Managing/Client/Object/ObjectCaching.cs:214)
FishNet.Managing.Client.ClientObjects:IterateObjectCache () (at Assets/FishNet/Runtime/Managing/Client/Object/ClientObjects.cs:386)
FishNet.Managing.Client.ClientManager:ParseReceived (FishNet.Transporting.ClientReceivedDataArgs) (at Assets/FishNet/Runtime/Managing/Client/ClientManager.cs:396)
FishNet.Managing.Client.ClientManager:Transport_OnClientReceivedData (FishNet.Transporting.ClientReceivedDataArgs) (at Assets/FishNet/Runtime/Managing/Client/ClientManager.cs:231)
FishNet.Transporting.Tugboat.Tugboat:HandleClientReceivedDataArgs (FishNet.Transporting.ClientReceivedDataArgs) (at Assets/FishNet/Runtime/Transporting/Transports/Tugboat/Tugboat.cs:217)
FishNet.Transporting.Tugboat.Client.ClientSocket:IterateIncoming () (at Assets/FishNet/Runtime/Transporting/Transports/Tugboat/Core/ClientSocket.cs:288)
FishNet.Transporting.Tugboat.Tugboat:IterateIncoming (bool) (at Assets/FishNet/Runtime/Transporting/Transports/Tugboat/Tugboat.cs:190)
FishNet.Managing.Transporting.TransportManager:IterateIncoming (bool) (at Assets/FishNet/Runtime/Managing/Transporting/TransportManager.cs:364)
FishNet.Managing.Timing.TimeManager:TryIterateData (bool) (at Assets/FishNet/Runtime/Managing/Timing/TimeManager.cs:955)
FishNet.Managing.Timing.TimeManager:IncreaseTick () (at Assets/FishNet/Runtime/Managing/Timing/TimeManager.cs:687)
FishNet.Managing.Timing.TimeManager:TickUpdate () (at Assets/FishNet/Runtime/Managing/Timing/TimeManager.cs:361)
FishNet.Transporting.NetworkReaderLoop:Update () (at Assets/FishNet/Runtime/Transporting/NetworkReaderLoop.cs:28)
```

Here are the scripts used:

```
using FishNet.Component.Observing;
using FishNet.Object;
using UnityEngine;

public class TestPlayer : NetworkBehaviour
{
    [SerializeField] GameObject objctToSpawn;
    [SerializeField] int match = 1;

    public override void OnStartServer()
    {
        base.OnStartServer();

        MatchCondition.AddToMatch(match, Owner);
    }

    void Update()
    {
        if (!IsServer) return;

        if (Input.GetKey(KeyCode.S))
        {
            GameObject go = Instantiate(objctToSpawn);
            Spawn(go);
        }

        if (Input.GetKeyDown(KeyCode.A))
        {
            match = match == 0 ? 1 : 0;
            MatchCondition.AddToMatch(match, Owner, replaceMatch: true);
        }
    }
}

```
```
using FishNet.Component.Observing;
using FishNet.Object;
using UnityEngine;

public class SpawnedObject : NetworkBehaviour
{
    [SerializeField] int lifespan = 60;
    int ticks = 0;

    public override void OnStartServer()
    {
        base.OnStartServer();

        MatchCondition.AddToMatch(0, NetworkObject);
    }

    void FixedUpdate()
    {
        if (!IsServer) return;
        
        ticks++;
        if (ticks > lifespan)
            Despawn();
    }
}
```

**Comment by FirstGearGames** [comment]
Thank you for the report. I'll be exploring this rather soon.

**Comment by FirstGearGames** [comment]
Unable to reproduce in 2.5.0, which will be out today. If 2.5 solves your problem please close this. Thanks :)

**Comment by maxkratt** [comment]
Problem is indeed resolved with version 2.5.0. Thank you!
